### PR TITLE
GH-107465: Add `pathlib.Path.from_uri()` classmethod.

### PR DESCRIPTION
**Pull request by barneygale**
This method supports file URIs (including variants) as described in RFC 8089, such as URIs generated by `pathlib.Path.as_uri()` and `urllib.request.pathname2url()`.

The method is added to `Path` rather than `PurePath` because it uses `os.fsdecode()`, and so its results vary from system to system. I intend to deprecate `PurePath.as_uri()` and move it to `Path` for the same reason.

* Issue: gh-107465

----
:books: Documentation preview :books:: https://cpython-previews--107640.org.readthedocs.build/

### CODE REVIEW

File: Lib/test/test_pathlib.py [hunk 1]
```diff
@@ -2913,6 +2914,20 @@ def test_passing_kwargs_deprecated(self):
         with self.assertWarns(DeprecationWarning):
             self.cls(foo="bar")
 
+    def test_from_uri_common(self):
+        P = self.cls
+        self.assertEqual(P.from_uri('file:foo/bar'), P('foo/bar'))
```
Comment (ap--) [inline Lib/test/test_pathlib.py]: Isn't this incorrect behaviour?

```pycon
# python3.11
>>> import pathlib
>>> pathlib.Path("foo/bar").as_uri()
ValueError: relative path can't be expressed as a file URI
```

Comment (barneygale) [reply Lib/test/test_pathlib.py]: I'm attempting to follow the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)

Comment (barneygale) [reply Lib/test/test_pathlib.py]: ... but the robustness principle gets it share of criticism too. My mind isn't set!

Another thing worth considering: all other pathlib constructors accept all string inputs without raising exceptions. We've previously discussed and decided against performing validation, e.g. checking for control characters in Windows paths. Whether or not that logic extends here is another matter.

Comment (ap--) [reply Lib/test/test_pathlib.py]: Oh, I don't have a strong preference either way. It's more that: whatever pathlib decides to do here will influence what universal_pathlib will support. Like in the [workflow xkcd](https://xkcd.com/1172/), someone will rely on the behaviour 😄 

In support for relative file uris: people seem to expect it to work (see here https://github.com/fsspec/universal_pathlib/commit/662c2943deff680423e048f62326cfdb8405ef09#r123906140 for example).

File: Lib/pathlib.py [hunk 2]
```diff
@@ -1184,6 +1184,25 @@ def __new__(cls, *args, **kwargs):
             cls = WindowsPath if os.name == 'nt' else PosixPath
         return object.__new__(cls)
 
+    @classmethod
+    def from_uri(cls, uri):
+        """Return a new path from the given 'file' URI."""
+        uri = uri.removeprefix('file:')
```
Comment (edgarrmondragon) [inline Lib/pathlib.py]: Should the method fail if the URI starts with something other than `file:`?

(looking at https://github.com/pypa/pip/blob/6f3a7181b6bfa0e95d479b3269baa0e551ff2cb2/src/pip/_internal/utils/urls.py#L26-L62)

Comment (barneygale) [reply Lib/pathlib.py]: Quite possibly.

The main `PurePath()` / `Path()` constructor does not perform any validation (for example, it doesn't reject paths with null characters). I figured we could make `from_uri()` work the same, but it's a pretty weak rationale.

File: Doc/library/pathlib.rst [hunk 3]
```diff
@@ -850,6 +850,49 @@ call fails (for example because the path doesn't exist).
...
+
+   URIs with no slash after the scheme (and no drive letter) are parsed as
+   relative paths::
+
+       >>> p = Path.from_uri('file:foo/bar')
+       WindowsPath('foo/bar')
+
+   Users may wish to test the result with :meth:`~PurePath.is_absolute` and
+   reject relative paths, as these are not portable across processes with
+   different working directories.
```
Comment (AA-Turner) [inline Doc/library/pathlib.rst]: I found this sentence a little hard to follow
```suggestion
   Users may wish to test the result of :meth:`!from_uri` with 
   :meth:`~PurePath.is_absolute`, as relative paths are not portable
   across processes with different working directories.
```

File: Doc/library/pathlib.rst [hunk 4]
```diff
@@ -850,6 +850,49 @@ call fails (for example because the path doesn't exist).
...
+   Several variant forms are supported::
+
+       >>> p = Path.from_uri('file:////server/share')
+       WindowsPath('//server/share')
+       >>> p = Path.from_uri('file://///server/share')
+       WindowsPath('//server/share')
+       >>> p = Path.from_uri('file:c:/windows')
+       WindowsPath('c:/windows')
+       >>> p = Path.from_uri('file:/c|/windows')
+       WindowsPath('c:/windows')
```
Comment (AA-Turner) [inline Doc/library/pathlib.rst:877]: I thought ``|`` was a banned character on Windows!

File: Doc/whatsnew/3.13.rst [hunk 5]
```diff
@@ -170,6 +170,9 @@ pathlib
   :exc:`NotImplementedError` when a path operation isn't supported.
   (Contributed by Barney Gale in :gh:`89812`.)
 
+* Add :meth:`pathlib.Path.from_uri` classmethod.
```
Comment (AA-Turner) [inline Doc/whatsnew/3.13.rst]: I think this needs an article:
```suggestion
* Add the :meth:`pathlib.Path.from_uri` classmethod.
```

File: Lib/pathlib.py [hunk 6]
```diff
@@ -814,9 +814,10 @@ class _PathBase(PurePath):
     __bytes__ = None
     __fspath__ = None  # virtual paths have no local file system representation
 
-    def _unsupported(self, method_name):
-        msg = f"{type(self).__name__}.{method_name}() is unsupported"
-        if isinstance(self, Path):
+    @classmethod
+    def _unsupported(cls, method_name):
+        msg = f"{cls.__name__}.{method_name}() is unsupported"
+        if issubclass(cls, Path):
```
Comment (AA-Turner) [inline Lib/pathlib.py:820]: Could this be it's own PR?

File: Doc/library/pathlib.rst [hunk 7]
```diff
@@ -850,6 +850,49 @@ call fails (for example because the path doesn't exist).
...
+
+   URIs with no slash after the scheme (and no drive letter) are parsed as
+   relative paths::
+
+       >>> p = Path.from_uri('file:foo/bar')
+       WindowsPath('foo/bar')
```
Comment (AA-Turner) [inline Doc/library/pathlib.rst]: Wikipedia (an authoritative reference...) [suggests](https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?) that URIs with the ``file`` scheme must always contain at least one slash:

> A valid file URI must therefore begin with either `file:/path` (no hostname), `file:///path` (empty hostname), or `file://hostname/path`.

Supposing that this is true, do we want to support non-standards-compliant behaviour?

File: Doc/library/pathlib.rst (hunk 4 above)
Comment (barneygale) [reply Doc/library/pathlib.rst:877]: [According to wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme#Windows_2):

> On Microsoft Windows systems, the normal colon (:) after a device letter has sometimes been replaced by a vertical bar (|) in file URLs. This reflected the original URL syntax, which made the colon a reserved character in a path part.

It's covered in [RFC 8089 appendix E.2.2](https://datatracker.ietf.org/doc/html/rfc8089#appendix-E.2.2).

File: Doc/whatsnew/3.13.rst (hunk 5 above)
Comment (barneygale) [reply Doc/whatsnew/3.13.rst]: Hm, that doesn't appear in-keeping with the way other new functions/methods are mentioned in the document. Would this work?

```suggestion
* Add :meth:`pathlib.Path.from_uri` classmethod that creates a path object from a 'file' URI.
```

File: Doc/library/pathlib.rst (hunk 7 above)
Comment (barneygale) [reply Doc/library/pathlib.rst]: It's arguable. All the file URI standards are post-hoc, and sometimes file URIs are supported alongside [relative URLs](https://datatracker.ietf.org/doc/html/rfc1808), but in that case the `file:` scheme prefix is usually forbidden. E.g. `man 7 uri` says:

> Note that if you really mean to say "start from the current location," don't specify the scheme at all; use a relative address like `../test.txt`, which has the side-effect of being scheme-independent.

[`urllib.request.url2pathname()`](https://docs.python.org/3/library/urllib.request.html#urllib.request.url2pathname) supports relative paths and doesn't require the `file:` prefix, but it's very old and somewhat hard to use correctly.

A universal_pathlib user requested support for relative URIs with a `file:` prefix [here](https://github.com/fsspec/universal_pathlib/commit/662c2943deff680423e048f62326cfdb8405ef09#r123906140).

We could perhaps add a *strict* argument that rejects URIs unless they begin `file:/`?

**Comment by barneygale** [comment]
Thanks for your feedback, both. I've revised the code: `Path.from_uri()` now raises `ValueError` if the URI does not begin with `file:`, or if it encodes a relative path.

We could add a non-strict mode, but I think that's best done in a separate PR that also considers `to_uri()`.

File: Doc/library/pathlib.rst (hunk 7 above)
Comment (AA-Turner) [reply Doc/library/pathlib.rst]: > We could perhaps add a _strict_ argument that rejects URIs unless they begin `file:/`?

I'd support this, defaulting to True.

File: Lib/pathlib.py (hunk 6 above)
Comment (barneygale) [reply Lib/pathlib.py:820]: Could you expand a bit? As a standalone PR it's not very well motivated: all the callers of `_unsupported` are instance methods, and so there's no reason to change it to a classmethod. Before I was a core dev I found it pretty impossible to land such PRs, as they're easily regarded as churn.

File: Doc/whatsnew/3.13.rst (hunk 5 above)
Comment (AA-Turner) [reply Doc/whatsnew/3.13.rst]: > Hm, that doesn't appear in-keeping with the way other new functions/methods are mentioned in the document

*What's New* will be edited quite a lot before October next year -- I think it's better to do both for better grammar and style (just think of your GCSE English teacher!)

Perhaps:

```suggestion
* Add :meth:`pathlib.Path.from_uri`, a new constructor to create a :class:`pathlib.Path`
  object from a 'file' URI (``file:/``).
```

File: Lib/pathlib.py (hunk 6 above)
Comment (AA-Turner) [reply Lib/pathlib.py:820]: On a first glance it seemed unrelated to the core function of this PR, but I see why you've included it in this PR now, so no objections.

File: Lib/test/test_pathlib.py [hunk 8]
```diff
@@ -3721,6 +3739,30 @@ def check():
...
+        self.assertEqual(P.from_uri('file:/c|/path/to/file'), P('c:/path/to/file'))
+        self.assertEqual(P.from_uri('file:///c|/path/to/file'), P('c:/path/to/file'))
+        # UNC paths
+        self.assertEqual(P.from_uri('file://server/path/to/file'), P('//server/path/to/file'))
+        self.assertEqual(P.from_uri('file:////server/path/to/file'), P('//server/path/to/file'))
+        self.assertEqual(P.from_uri('file://///server/path/to/file'), P('//server/path/to/file'))
+        # Localhost paths
+        self.assertEqual(P.from_uri('file://localhost/c:/path/to/file'), P('c:/path/to/file'))
+        self.assertEqual(P.from_uri('file://localhost/c|/path/to/file'), P('c:/path/to/file'))
+        # Invalid paths
```
Comment (AA-Turner) [inline Lib/test/test_pathlib.py:3757]: Maybe a test for a URI with scheme but non ``file``?

**Review by AA-Turner — APPROVED** [review]
Thanks!

A

File: Doc/library/pathlib.rst (hunk 7 above)
Comment (barneygale) [reply Doc/library/pathlib.rst]: I've made the implementation stricter. I'll look at adding a *non*-strict mode later, if there's demand.

**Comment by barneygale** [comment]
Thank you for the reviews, Adam, Andreas and Edgar!
